tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,146 @@
+# limitations under the License.
+# ==============================================================================
+"""Discriminative Layer Training Optimizer for TensorFlow."""
+
+import tensorflow as tf
+from typeguard import typechecked
+from typing import Union
+from tensorflow.keras.optimizers import Optimizer
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class MultiOptimzer(Optimizer):
+    @typechecked
+    def __init__(
+        self,
+        optimizer_layer_pairs: Union[list, None] = None,
+        optimizer_specs: Union[list, None] = None,
+        name: str = "MultiOptimzer",
+        **kwargs
+    ):
+
+        """
+        Creates a wrapper around a set of instantiated optimizer layer pairs. Allows for implementation of
+        Discriminative Layer Training as per https://arxiv.org/abs/1607.05440. Generally useful for transfer learning
+        of deep networks.
+
+        Each optimizer will optimize only the weights associated with its paired layer. This can be used
+        to implement discriminative layer training by assigning different learning rates to each optimizer
+        layer pair. (Optimizer, list(Layers)) pairs are also supported. Please note that the layers must be
+        instantiated before instantiating the optimizer.
+
+        Currently, MultiOpt does not support callbacks that modify optimizers. However, you can instantiate
+        optimizer layer pairs with tf.keras.optimizers.schedules.LearningRateSchedule instead of a static learning
+        rate.
+
+        This code should function on CPU, GPU, and TPU. Apply the with strategy.scope() context as you
+        would with any other optimizer.
+
+        Example:
+
+        ```python
+
+        model = get_model()
+
+        opt1 = tf.keras.optimizers.Adam(learning_rate=1e-4)
+        opt2 = tf.keras.optimizers.Adam(learning_rate=1e-2)
+
+        opt_layer_pairs = [(opt1, model.layers[0]),
+                            (opt2, model.layers[1:])]
+
+        loss = tf.keras.losses.MSE
+        optimizer = tfa.optimizers.MultiOpt(opt_layer_pairs)
+
+        model.compile(optimizer=optimizer, loss = loss)
+
+        model.fit(x,y)
+
+        ```
+
+
+        """
+
+        super(MultiOptimzer, self).__init__(name, **kwargs)
+
+        if optimizer_specs is None and optimizer_layer_pairs is not None:
+            self.optimizer_specs = [
+                self.create_optimizer_spec(opt, layer)
+                for opt, layer in optimizer_layer_pairs
+            ]
+
+        elif optimizer_specs is not None and optimizer_layer_pairs is None:
+            self.optimizer_specs = [
+                self.maybe_initialize_optimizer_spec(spec) for spec in optimizer_specs
+            ]
+
+        else:
+            raise RuntimeError(
+                "You must specify either an list of optimizer_layer_pairs or a list of optimizer_specs"
+            )
+
+    def apply_gradients(self, grads_and_vars, name=None):
```

Comment:
This should be
```python
    def apply_gradients(self, grads_and_vars, **kwargs):
        ...
        return tf.group(
            [
                spec["optimizer"].apply_gradients(spec["gv"], **kwargs)
                for spec in self.optimizer_specs
            ]
        )
```
to prevent upstream change.